File: crates/rqlv2/src/tokenize/keyword.rs
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (c) 2025 ReifyDB

//! Keyword definitions.

use std::{collections::HashMap, sync::LazyLock};

macro_rules! keyword {
    ( $( $variant:ident => $string:literal ),* $(,)? ) => {
        /// RQL keywords.
        #[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
        pub enum Keyword {
            $( $variant ),*
        }

        impl Keyword {
            /// Get the canonical string representation.
            pub const fn as_str(&self) -> &'static str {
                match self {
                    $( Keyword::$variant => $string ),*
                }
            }
        }

        /// Map from uppercase keyword strings to Keyword variants.
        pub static KEYWORD_MAP: LazyLock<HashMap<&'static str, Keyword>> = LazyLock::new(|| {
            let mut map = HashMap::new();
            $( map.insert($string, Keyword::$variant); )*
            map
        });
    };
}

// 65 keywords (all case-insensitive)
keyword! {
    // Query clauses
    Map        => "MAP",
    Select     => "SELECT",
    Extend     => "EXTEND",
    By         => "BY",
    From       => "FROM",
    Where      => "WHERE",
    Aggregate  => "AGGREGATE",
    Sort       => "SORT",
    Distinct   => "DISTINCT",
    Take       => "TAKE",
    Offset     => "OFFSET",
    Filter     => "FILTER",
    Scan       => "SCAN",

    // Joins
    Left       => "LEFT",
    Inner      => "INNER",
    Natural    => "NATURAL",
    Join       => "JOIN",
    On         => "ON",
    Using      => "USING",
    Merge      => "MERGE",

    // DML
    Insert     => "INSERT",
    Into       => "INTO",
    Update     => "UPDATE",
    Set        => "SET",
    Delete     => "DELETE",

    // Control flow
    Let        => "LET",
    Mut        => "MUT",
    If         => "IF",
    Else       => "ELSE",
    End        => "END",
    Loop       => "LOOP",
    Return     => "RETURN",
    For        => "FOR",

    // Functions
    Define     => "DEFINE",
    Function   => "FUNCTION",
    Call       => "CALL",
    Apply      => "APPLY",
    Cast       => "CAST",

    // DDL
    Describe   => "DESCRIBE",
    Show       => "SHOW",
    Create     => "CREATE",
    Alter      => "ALTER",
    Drop       => "DROP",
    Rename     => "RENAME",

    // Schema objects
    Namespace  => "NAMESPACE",
    Sequence   => "SEQUENCE",
    Series     => "SERIES",
    Table      => "TABLE",
    Ringbuffer => "RINGBUFFER",
    Policy     => "POLICY",
    View       => "VIEW",
    Flow       => "FLOW",
    Window     => "WINDOW",
    Dictionary => "DICTIONARY",
    Index      => "INDEX",
    Query      => "QUERY",

    // Constraints
    Unique     => "UNIQUE",
    Primary    => "PRIMARY",
    Key        => "KEY",

    // Modifiers
    Deferred      => "DEFERRED",
    Transactional => "TRANSACTIONAL",
    Asc           => "ASC",
    Desc          => "DESC",
    Auto          => "AUTO",
    Increment     => "INCREMENT",
    Value         => "VALUE",

    // Existence
    Exists     => "EXISTS",
    Replace    => "REPLACE",
    Cascade    => "CASCADE",
    Restrict   => "RESTRICT",

    // Misc
    In         => "IN",
    Between    => "BETWEEN",
    Like       => "LIKE",
    Is         => "IS",
    With       => "WITH",
    To         => "TO",
    Pause      => "PAUSE",
    Resume     => "RESUME",
    Rownum     => "ROWNUM",
}

/// Try to match an identifier string to a keyword (case-insensitive).
pub fn lookup_keyword(s: &str) -> Option<Keyword> {
	// Fast path for already uppercase
	if s.chars().all(|c| c.is_ascii_uppercase()) {
		KEYWORD_MAP.get(s).copied()
	} else {
		// Convert to uppercase for lookup
		let upper = s.to_ascii_uppercase();
		KEYWORD_MAP.get(upper.as_str()).copied()
	}
}

#[cfg(test)]
mod tests {
	use super::*;

	#[test]
	fn test_keyword_lookup_uppercase() {
		assert_eq!(lookup_keyword("SELECT"), Some(Keyword::Select));
		assert_eq!(lookup_keyword("MAP"), Some(Keyword::Map));
		assert_eq!(lookup_keyword("SCAN"), Some(Keyword::Scan));
	}

	#[test]
	fn test_keyword_lookup_lowercase() {
		assert_eq!(lookup_keyword("select"), Some(Keyword::Select));
		assert_eq!(lookup_keyword("map"), Some(Keyword::Map));
		assert_eq!(lookup_keyword("scan"), Some(Keyword::Scan));
	}

	#[test]
	fn test_keyword_lookup_mixed_case() {
		assert_eq!(lookup_keyword("Select"), Some(Keyword::Select));
		assert_eq!(lookup_keyword("MaP"), Some(Keyword::Map));
		assert_eq!(lookup_keyword("ScAn"), Some(Keyword::Scan));
	}

	#[test]
	fn test_keyword_lookup_not_found() {
		assert_eq!(lookup_keyword("foobar"), None);
		assert_eq!(lookup_keyword("SELECTX"), None);
	}

	#[test]
	fn test_keyword_as_str() {
		assert_eq!(Keyword::Select.as_str(), "SELECT");
		assert_eq!(Keyword::Map.as_str(), "MAP");
		assert_eq!(Keyword::Scan.as_str(), "SCAN");
	}
}

File: pkg/typescript/core/src/syntax/keyword.ts
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) 2026 ReifyDB

const QUERY_TRANSFORMS = [
    'map', 'extend', 'by', 'from', 'where', 'aggregate', 'having',
    'sort', 'distinct', 'take', 'offset',
];

const JOINS_AND_SET_OPERATIONS = [
    'left', 'inner', 'natural', 'join', 'on', 'using', 'intersect', 'except',
];

const DML = ['insert', 'into', 'update', 'set', 'delete'];

const CONTROL_FLOW = [
    'let', 'if', 'else', 'end', 'loop', 'while', 'break', 'continue', 'return',
];

const FUNCTIONS_AND_CASTING = ['fun', 'call', 'apply', 'cast'];

const DDL_AND_SHAPE = [
    'describe', 'show', 'create', 'alter', 'drop', 'filter', 'gate', 'flow', 'window',
    'returning',
];

const OPERATOR_KEYWORDS = ['in', 'between', 'like', 'is', 'with'];

const OBJECT_TYPES = [
    'namespace', 'sequence', 'series', 'subscription', 'table', 'ringbuffer',
    'column', 'policy', 'property', 'view', 'deferred', 'transactional',
];

const INDEX_AND_CONSTRAINTS = [
    'index', 'unique', 'primary', 'key', 'asc', 'desc', 'auto', 'increment', 'value',
];

const MISC_OPERATIONS = [
    'exists', 'replace', 'cascade', 'restrict', 'to', 'pause', 'resume',
    'query', 'rename', 'rownum', 'dictionary', 'for', 'output', 'append',
    'assert', 'patch',
];

const ENUMS_AND_PATTERN_MATCHING = ['enum', 'match'];

const PROCEDURES_AND_EVENTS = ['procedure', 'event', 'handler', 'dispatch', 'tag'];

const TESTING = ['test', 'tests', 'run'];

const ACCESS_CONTROL = [
    'user', 'role', 'grant', 'revoke', 'password', 'require', 'execute',
    'access', 'subscribe', 'enable', 'disable',
];

const SYSTEM_OBJECTS = ['function', 'session', 'feature'];

const MIGRATIONS = [
    'add', 'migration', 'migrate', 'rollback', 'diff', 'version', 'current', 'pending',
];

const MISC = ['authentication', 'contains', 'remote', 'error'];

const ADDITIONAL_CONSTRUCTS = ['derive', 'group', 'union', 'as'];

const WORD_OPERATORS = ['and', 'or', 'not', 'xor'];

export const RQL_KEYWORDS: readonly string[] = [
    ...QUERY_TRANSFORMS,
    ...JOINS_AND_SET_OPERATIONS,
    ...DML,
    ...CONTROL_FLOW,
    ...FUNCTIONS_AND_CASTING,
    ...DDL_AND_SHAPE,
    ...OPERATOR_KEYWORDS,
    ...OBJECT_TYPES,
    ...INDEX_AND_CONSTRAINTS,
    ...MISC_OPERATIONS,
    ...ENUMS_AND_PATTERN_MATCHING,
    ...PROCEDURES_AND_EVENTS,
    ...TESTING,
    ...ACCESS_CONTROL,
    ...SYSTEM_OBJECTS,
    ...MIGRATIONS,
    ...MISC,
    ...ADDITIONAL_CONSTRUCTS,
    ...WORD_OPERATORS,
];

export const RQL_LITERALS: readonly string[] = ['none', 'true', 'false'];

const SIGNED_INTEGERS = ['int1', 'int2', 'int4', 'int8', 'int16', 'int'];

const UNSIGNED_INTEGERS = ['uint1', 'uint2', 'uint4', 'uint8', 'uint16', 'uint'];

const FLOATING_POINT = ['float4', 'float8'];

const TEXT_AND_BINARY = ['utf8', 'blob', 'text'];

const BOOLEAN = ['bool', 'boolean'];

const NUMERIC = ['decimal'];

const TEMPORAL = ['date', 'datetime', 'time', 'duration', 'interval'];

const IDENTIFIERS_AND_UUIDS = [
    'uuid4', 'uuid7', 'identityid', 'identity_id', 'dictionaryid', 'dictionary_id',
];

const CONTAINERS = ['Option', 'List', 'Record', 'Tuple'];

export const RQL_TYPES: readonly string[] = [
    ...SIGNED_INTEGERS,
    ...UNSIGNED_INTEGERS,
    ...FLOATING_POINT,
    ...TEXT_AND_BINARY,
    ...BOOLEAN,
    ...NUMERIC,
    ...TEMPORAL,
    ...IDENTIFIERS_AND_UUIDS,
    ...CONTAINERS,
    'any',
];

export const RQL_OPERATORS: readonly string[] = [
    '+', '-', '*', '/', '//', '%', '=', '==', '!=', '->', '=>', '>', '<', '>=', '<=',
    '~=', '&&', '||', '??',
];

export const RQL_SYSTEM_COLUMNS: readonly string[] = ['rownum', 'created_at', 'updated_at'];
